### PR TITLE
Scoped implicitly-typed lambda parameters

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/62080
Corresponding spec update: https://github.com/dotnet/csharplang/pull/6551

Relates to test plan https://github.com/dotnet/roslyn/issues/59194 (ref fields)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 1]
```diff
@@ -611,7 +613,8 @@ private static TypeWithAnnotations DelegateReturnTypeWithAnnotations(MethodSymbo
             }
 
             var parameterRefKindsBuilder = ArrayBuilder<RefKind>.GetInstance(ParameterCount);
-            var parameterScopesBuilder = ArrayBuilder<DeclarationScope>.GetInstance(ParameterCount);
+            var parameterDeclarationScopesBuilder = ArrayBuilder<DeclarationScope>.GetInstance(ParameterCount);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:616]: >parameterDeclarationScopesBuilder [](http://example.com/codeflow?start=16&length=33)

Consider calling this `parameterDeclaredScopesBuilder`.

File: src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs [hunk 2]
```diff
@@ -304,9 +307,11 @@ internal override void AddDeclarationDiagnostics(BindingDiagnosticBag diagnostic
             CSharpCompilation compilation,
             UnboundLambda unboundLambda,
             ImmutableArray<TypeWithAnnotations> parameterTypes,
-            ImmutableArray<RefKind> parameterRefKinds)
+            ImmutableArray<RefKind> parameterRefKinds,
+            ImmutableArray<DeclarationScope> parameterDeclarationScopes)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs:311]: >parameterDeclarationScopes [](http://example.com/codeflow?start=45&length=26)

Consider calling `parameterDeclaredScopes`.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs [hunk 3]
```diff
@@ -225,5 +225,28 @@ internal static bool IsValidUnscopedRefAttributeTarget(this MethodSymbol method)
                 !method.IsConstructor() &&
                 !method.IsInitOnly;
         }
+
+#nullable enable
+        internal static ImmutableArray<DeclarationScope> GetByValueParameterDeclarationScopes(this MethodSymbol? method)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs:230]: >GetByValueParameterDeclarationScopes [](http://example.com/codeflow?start=57&length=36)

I think it would be helpful if these methods are named as "...DeclaredScopes" or "...EffectiveScopes". Currently, it's not clear if "...DeclarationScopes" refers to the enum type name or the declared values.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 4]
```diff
@@ -1054,6 +1094,7 @@ public static ReturnInferenceCacheKey Create(NamedTypeSymbol? delegateType, bool
 
                         parameterTypes = typesBuilder.ToImmutableAndFree();
                         parameterRefKinds = refKindsBuilder.ToImmutableAndFree();
+                        parameterDeclarationScopes = MethodSymbolExtensions.GetByValueParameterDeclarationScopes(invoke);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:1097]: >MethodSymbolExtensions.GetByValueParameterDeclarationScopes(invoke) [](http://example.com/codeflow?start=53&length=67)

`invoke.GetByValueParameterDeclarationScopes()`?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 5]
```diff
@@ -1002,6 +1020,24 @@ public override bool Equals(object? obj)
...
+                    return false;
+                }
+
+                for (int i = 0; i < this.ParameterDeclarationScopes.Length; i++)
+                {
+                    if (this.ParameterDeclarationScopes[i] != other.ParameterDeclarationScopes[i])
+                    {
+                        return false;
+                    }
+                }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:1039]: `return ParameterDeclarationScopes.SequenceEqual(other.ParameterDeclarationScopes);`

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 6]
```diff
@@ -12458,6 +12457,221 @@ static void New()
...
+                    public ref int field;
+                }
+
+                class C
+                {
+                    void M()
+                    {
+                        M2(r => r);
+                        M2((scoped R r) => r); // 1
+                    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12509]: >} [](http://example.com/codeflow?start=20&length=1)

Consider testing `M2(r => default);`. Should overload resolution fail in that case, with ambiguous overloads?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 7]
```diff
@@ -12458,6 +12457,221 @@ static void New()
...
+                public ref struct R
+                {
+                    public ref int field;
+                }
+
+                class C
+                {
+                    void M()
+                    {
+                        D1 d1 = () => r;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12549]: >r [](http://example.com/codeflow?start=38&length=1)

Consider changing to `new R()`, to avoid any impact from that error.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 8]
```diff
@@ -12458,6 +12457,221 @@ static void New()
...
+                {
+                    public ref int field;
+                }
+
+                class C
+                {
+                    void M()
+                    {
+                        D1 d1 = (scoped R r) => r;
+                    }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:12577]: >} [](http://example.com/codeflow?start=20&length=1)

Consider testing `d1 = r => r;` as well.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 9]
```diff
@@ -1002,7 +1020,17 @@ public override bool Equals(object? obj)
                     }
                 }
 
-                return true;
+                if (this.ParameterDeclarationScopes.IsDefault || other.ParameterDeclarationScopes.IsDefault)
+                {
+                    return this.ParameterDeclarationScopes.IsDefault == other.ParameterDeclarationScopes.IsDefault;
+                }
+
+                if (this.ParameterDeclarationScopes.Length != other.ParameterDeclarationScopes.Length)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:1028]: >if (this.ParameterDeclarationScopes.Length != other.ParameterDeclarationScopes.Length) [](http://example.com/codeflow?start=16&length=86)

Minor: I believe this is included in `SequenceEqual()` below.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 10]
```diff
@@ -964,16 +976,22 @@ private sealed class ReturnInferenceCacheKey
...
-            public static readonly ReturnInferenceCacheKey Empty = new ReturnInferenceCacheKey(ImmutableArray<TypeWithAnnotations>.Empty, ImmutableArray<RefKind>.Empty, null);
+            public static readonly ReturnInferenceCacheKey Empty = new ReturnInferenceCacheKey(ImmutableArray<TypeWithAnnotations>.Empty, ImmutableArray<RefKind>.Empty,
+                parameterDeclaredScopes: default, taskLikeReturnTypeOpt: null);
 
-            private ReturnInferenceCacheKey(ImmutableArray<TypeWithAnnotations> parameterTypes, ImmutableArray<RefKind> parameterRefKinds, NamedTypeSymbol? taskLikeReturnTypeOpt)
+            private ReturnInferenceCacheKey(ImmutableArray<TypeWithAnnotations> parameterTypes, ImmutableArray<RefKind> parameterRefKinds,
+                ImmutableArray<DeclarationScope> parameterDeclaredScopes, NamedTypeSymbol? taskLikeReturnTypeOpt)
             {
                 Debug.Assert(parameterTypes.Length == parameterRefKinds.Length);
+                Debug.Assert(parameterDeclaredScopes.IsDefault || parameterTypes.Length == parameterDeclaredScopes.Length);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:989]: >parameterDeclaredScopes.IsDefault [](http://example.com/codeflow?start=29&length=33)

What is the scenario when this condition is true?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:989]: I'll add today.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:989]: When all parameters are unscoped, we use a `default` array.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs [hunk 11]
```diff
@@ -225,5 +225,28 @@ internal static bool IsValidUnscopedRefAttributeTarget(this MethodSymbol method)
...
+                return default;
+
+            if (method.Parameters.All(p => getByValueParameterDeclaredScope(p) == DeclarationScope.Unscoped))
+                return default;
+
+            return method.Parameters.SelectAsArray(p => getByValueParameterDeclaredScope(p));
+
+            static DeclarationScope getByValueParameterDeclaredScope(ParameterSymbol parameter)
+            {
+                // For implicitly-typed lambda parameters, we only care about declaration scopes on by-value parameters.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs:242]: >For implicitly-typed lambda parameters, we only care about declaration scopes on by-value parameters. [](http://example.com/codeflow?start=19&length=101)

It is not obvious why is that.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs:242]: If a parameter is `ref` or `out` then the parameters must have been explicitly-typed.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs:242]: > If a parameter is ref or out then the parameters must have been explicitly-typed.

I am not sure I am following. We are looking at a method symbol. All parameters are explicitly typed for it.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs:242]: Still looking at other changes, I might get it at the end

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:989]: Created corresponding spec update: https://github.com/dotnet/csharplang/pull/6551

**Comment by jaredpar** [comment]
@roslyn-compiler please ignore this message. Testing out a GitHub rule change.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 12]
```diff
@@ -1502,7 +1502,7 @@ internal sealed class SourceComplexParameterSymbolWithCustomModifiersPrecedingRe
             SyntaxReference syntaxRef,
             bool isParams,
             bool isExtensionMethodThis,
-            DeclarationScope scope)
+            DeclarationScope? scope)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:1505]: >DeclarationScope? scope [](http://example.com/codeflow?start=12&length=23)

Is this change needed?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs [hunk 13]
```diff
@@ -1479,7 +1479,7 @@ internal sealed class SourceComplexParameterSymbol : SourceComplexParameterSymbo
...
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:1482]: >DeclarationScope? scope [](http://example.com/codeflow?start=12&length=23)

Is this change needed? Can we get here from a `LambdaParameterSymbol`?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:1482]: `WithCustomModifiersAndParamsCore` copies `this.DeclaredSymbol` (which is `Nullable<DeclarationScope>`) using this constructor.
It similarly calls the `SourceComplexParameterSymbolWithCustomModifiersPrecedingRef` constructor.

Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:1482]: Is `WithCustomModifiersAndParamsCore()` used for `LambdaParameterSymbol`? Otherwise, we'd presumably need to copy `EffectiveScope` as well.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs:1482]: `WithCustomModifiersAndParamsCore` isn't used for lambdas. It's used for OHI scenarios.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceParameterSymbol.cs [hunk 14]
```diff
@@ -225,13 +225,16 @@ public sealed override RefKind RefKind
...
+        internal DeclarationScope? DeclaredScope => _scope;
 
         internal abstract override DeclarationScope EffectiveScope { get; }
 
         protected DeclarationScope CalculateEffectiveScopeIgnoringAttributes()
         {
-            var declaredScope = this.DeclaredScope;
+            // DeclaredScope is only null in LambdaParameterSymbol where EffectiveScope can be provided rather than computed
+            Debug.Assert(this.DeclaredScope is not null);
+            var declaredScope = this.DeclaredScope.Value;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceParameterSymbol.cs:236]: >Value [](http://example.com/codeflow?start=51&length=5)

```ValueOrDefault```? Or is it your intent to crash on ```null``` value here? If so, perhaps the assert is redundant.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 9)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceParameterSymbol.cs (hunk 14 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceParameterSymbol.cs:236]: I'd rather leave as-is. I thought this was nullable-enabled code.
